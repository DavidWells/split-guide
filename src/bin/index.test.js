/* eslint no-console:0 */
import path from 'path'
import fs from 'fs'
import spawn from 'spawn-command'
import pify from 'pify'
import glob from 'glob'
import dirTree from 'directory-tree'

const SPLIT_GUIDE_PATH = require.resolve('./index')
const BABEL_BIN_PATH = require.resolve('babel-cli/bin/babel-node')

test('outputs helpful output', () => {
  return runSplitGuideCLI('--help').then(stdout => {
    expect(stdout).toMatchSnapshot()
  })
})

test('generates split-guides', () => {
  const generateFixtures = path.resolve(__dirname, '../../test/fixtures/generate')
  return runSplitGuideCLI('generate', generateFixtures).then(stdout => {
    expect(stdout).toMatchSnapshot()
    const tree = dirTree(generateFixtures)
    expect(tree).toMatchSnapshot()
    // cannot use Promise.all here because we need to make sure the snapshots are
    // taken in the correct order
    return expectDirectoryToMatchSnapshot(path.resolve(generateFixtures, './exercises'))
      .then(() => expectDirectoryToMatchSnapshot(path.resolve(generateFixtures, './exercises-final')))
  })
})

function expectDirectoryToMatchSnapshot(directory) {
  return pify(glob)(path.resolve(directory, '**/*'), {nodir: true})
    .then(readAllFilesAsPromise)
    .then(expectFilesToMatchSnapshot)

  function readAllFilesAsPromise(files) {
    const allPromises = files.map(readFileAsPromise)
    return Promise.all(allPromises)
  }

  function readFileAsPromise(file) {
    return pify(fs.readFile)(file, 'utf8').then(contents => ({file, contents}))
  }

  function expectFilesToMatchSnapshot(files) {
    expect(files).toMatchSnapshot()
  }
}

function runSplitGuideCLI(args = '', cwd = process.cwd()) {
  const isRelative = cwd[0] !== '/'
  if (isRelative) {
    cwd = path.resolve(__dirname, cwd)
  }

  return new Promise((resolve, reject) => {
    let stdout = ''
    let stderr = ''
    const command = `${BABEL_BIN_PATH} ${SPLIT_GUIDE_PATH} ${args}`
    const child = spawn(command, {cwd})

    child.on('error', error => {
      reject(error)
    })

    child.stdout.on('data', data => {
      stdout += data.toString()
    })

    child.stderr.on('data', data => {
      stderr += data.toString()
    })

    child.on('close', () => {
      if (stderr) {
        reject(stderr)
      } else {
        resolve(stdout)
      }
    })
  })
}
