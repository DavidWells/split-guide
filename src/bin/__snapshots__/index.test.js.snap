exports[`test generates split-guides 1`] = `""`;

exports[`test generates split-guides 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises/foo/bar/README.md",
                  "size": 65
                }
              ],
              "name": "bar",
              "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises/foo/bar",
              "size": 65
            }
          ],
          "name": "foo",
          "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises/foo",
          "size": 65
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises/root.js",
          "size": 92
        }
      ],
      "name": "exercises",
      "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises",
      "size": 157
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final/foo/bar/README.md",
                  "size": 61
                }
              ],
              "name": "bar",
              "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final/foo/bar",
              "size": 61
            }
          ],
          "name": "foo",
          "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final/foo",
          "size": 61
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final/root.js",
          "size": 76
        }
      ],
      "name": "exercises-final",
      "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final",
      "size": 137
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "extension": ".md",
                  "name": "README.md",
                  "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/templates/foo/bar/README.md",
                  "size": 292
                }
              ],
              "name": "bar",
              "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/templates/foo/bar",
              "size": 292
            }
          ],
          "name": "foo",
          "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/templates/foo",
          "size": 292
        },
        Object {
          "extension": ".js",
          "name": "root.js",
          "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/templates/root.js",
          "size": 261
        }
      ],
      "name": "templates",
      "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/templates",
      "size": 553
    }
  ],
  "name": "generate",
  "path": "/Users/kdodds/Developer/split-guide/test/fixtures/generate",
  "size": 847
}
`;

exports[`test generates split-guides 3`] = `
Array [
  Object {
    "contents": "This is a fixture

This is workshop stuff

This is common stuff

",
    "file": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises/foo/bar/README.md"
  },
  Object {
    "contents": "/* eslint-disable */
module.exports = function sum(a, b) {
  // return the sum of a and b
}
",
    "file": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises/root.js"
  }
]
`;

exports[`test generates split-guides 4`] = `
Array [
  Object {
    "contents": "This is a fixture

This is common stuff

This is final stuff
",
    "file": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final/foo/bar/README.md"
  },
  Object {
    "contents": "/* eslint-disable */
module.exports = function sum(a, b) {
  return a + b
}
",
    "file": "/Users/kdodds/Developer/split-guide/test/fixtures/generate/exercises-final/root.js"
  }
]
`;

exports[`test outputs helpful output 1`] = `
"Usage: src/bin/index.js

Commands:
  generate  generate your split-guide

Options:
  -h, --help  Show help  [boolean]

"
`;
